### PR TITLE
Core/Movement: Implement MoveSplineFlag::Enter_Cycle flag

### PR DESCRIPTION
**Pull request by Faq**
by xvwyh

**Changes proposed:**

-  
-  
-  

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #  (insert issue tracker number)
#22448
#20238

**Tests performed:** (Does it build, tested in-game, etc.)
Some testing was performed by Ovahlord described in #22448

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by Ovahlord** [comment]
Tested this stuff for a long time now, no issues to complain about.

**Comment by jackpoz** [comment]
@Ovahlord could you add the changes that Shauren mentioned in irc ?

**Comment by Faq** [comment]
Would be good to add here changes he mentioned on irc.

**Comment by jackpoz** [comment]
>[2020-01-14 20:09:45] <Shauren> eeeeeeeeh, im pretty sure you can do that calculation in a better way than initializing a duplicate spline
[2020-01-14 20:11:22] <Elon> u mean u can recalc the new spline duration and length after completing the first cycle without tearing half of the spline system apart?
[2020-01-14 20:13:09] <Shauren> totally
[2020-01-14 20:14:24] <Shauren> since you only need to drop first point, you basically just need to shorten the duration by the time it takes to move from first to second point
[2020-01-14 20:15:03] <Elon> does the spline system actually have a segment info?
[2020-01-14 20:15:11] <Elon> like duration per point?
[2020-01-14 20:15:46] <Shauren> which means you need to simply use spline.length(1, spline.last()) // where spline is the internal member of MoveSpline
[2020-01-14 20:16:03] <Shauren> that returns the duration with the first segment removed

**Comment by jackpoz** [comment]
@Faq any followup to the changes mentioned above ?

**Comment by Faq** [comment]
Sorry, no time. Anyone can do that.

**Comment by jackpoz** [comment]
One day someone will implement what Shauren suggested

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Comment by xvwyh** [comment]
Not sure that the thing @Shauren is suggesting is achievable as easily as he thinks. If we were talking about linear splines - absolutely. But with curved splines, removing a point can radically alter the curvature of surrounding segments, and thus their length. And even the angle of approach to the first point in the loop will be much different with and without the enter_cycle point, which will also affect the curvature and length. Unless Catmull-Rom splines have some magical properties that I don't know about.

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -484,19 +485,28 @@ void Unit::UpdateSplineMovement(uint32 t_diff)
...
+            data.appendPackGUID(GetGUID());
+            SendMessageToSet(&data, true);
+        }
+    }
+
     if (arrived)
         DisableSpline();
 
-    m_movesplineTimer.Update(t_diff);
-    if (m_movesplineTimer.Passed() || arrived)
```
Comment (jackpoz) [inline src/server/game/Entities/Unit/Unit.cpp]: could this timer be restored to avoid calling UpdateSplinePosition() every update ?

Comment (Ovahlord) [reply src/server/game/Entities/Unit/Unit.cpp:491]: It's less about updating the position, it's about the following triggered actions such as vmap based position data updates. Long story short, this change is not only updating spline positions but also causing vmap lookups which are now being spamed with every single update tick. Not really a healthy thing to do

Comment (xvwyh) [reply src/server/game/Entities/Unit/Unit.cpp:491]: Then maybe that's the thing you want to optimize instead? Throttle (or completely block) movement flags recalc for creatures who are using random or waypoint movement out of combat, throttle visibility updates and MoveInLineOfSight calls. IMO that's a better solution than causing server/client desyncs.

And as for dynamic spawns, I don't know how they work, but I cannot see why can zone be changeable in real time: blizzard used separate servers for separate zones from the beginning, as they themselves said, and I highly doubt that it makes sense for spawns to be able to transition from zone to zone under these circumstances, as that would mean they would have to be moved to a different server responsible for that zone, which is just a technical overkill.

Comment (Jildor) [reply src/server/game/Entities/Unit/Unit.cpp:491]: this can be related with this issue?
https://github.com/TrinityCore/TrinityCore/issues/24356

Comment (xvwyh) [reply src/server/game/Entities/Unit/Unit.cpp:491]: If the claim "seems related with pathfinding" is true - then no. Pathfinding would be executed in the same amounts with or without this position update delay. This delay is only responsible for controlling how often is the unit being physically moved along the already generated path.

**Pull request merged**